src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java

```diff
@@ -52,4 +94,304 @@ private long countGermplasm(UUID programId, String trialDbId) throws ApiExceptio
+        return new DownloadFile(fileName, downloadFile);
+    }
+
+    private void addBrAPIObsToRecords(
+            List<BrAPIObservation> dataset,
+            BrAPITrial experiment,
+            Program program,
+            Map<String, BrAPIObservationUnit> ouByOUDbId,
+            Map<String, BrAPIStudy> studyByDbId,
+            Map<String, Map<String, Object>> rowByOUId,
+            boolean includeTimestamp,
+            List<BrAPIObservationVariable> obsVars) throws ApiException, DoesNotExistException {
+        Map<String, BrAPIObservationVariable> varByDbId = new HashMap<>();
+        obsVars.forEach(var -> varByDbId.put(var.getObservationVariableDbId(), var));
+        for (BrAPIObservation obs: dataset) {
+
+            // get observation unit for observation
+            BrAPIObservationUnit ou = ouByOUDbId.get(obs.getObservationUnitDbId());
+            String ouId = getOUId(ou);
+
+            // get observation variable for BrAPI observation
+            BrAPIObservationVariable var = varByDbId.get(obs.getObservationVariableDbId());
+
+            // if there is a row with that ouId then just add the obs var data and timestamp to the row
+            if (rowByOUId.get(ouId) != null) {
+                addObsVarDataToRow(rowByOUId.get(ouId), obs, includeTimestamp, var, program);
+            } else {
+
+                // otherwise make a new row
+                Map<String, Object> row = createExportRow(experiment, program, ou, studyByDbId);
+                addObsVarDataToRow(row, obs, includeTimestamp, var, program);
+                rowByOUId.put(ouId, row);
+            }
+        }
+    }
+
+    private String getOUId(BrAPIObservationUnit ou) {
+        BrAPIExternalReference ouXref = Utilities.getExternalReference(
+                        ou.getExternalReferences(),
+                        String.format("%s/%s", referenceSource, ExternalReferenceSource.OBSERVATION_UNITS.getName()))
+                .orElseThrow(() -> new RuntimeException("observation unit id not found"));
+        return ouXref.getReferenceID();
+    }
+
+    private String getStudyId(BrAPIStudy study) {
+        BrAPIExternalReference studyXref = Utilities.getExternalReference(
+                        study.getExternalReferences(),
+                        String.format("%s/%s", referenceSource, ExternalReferenceSource.STUDIES.getName()))
+                .orElseThrow(() -> new RuntimeException("study id not found"));
+        return studyXref.getReferenceID();
+    }
+
+    private void addObsVarDataToRow(
+            Map<String, Object> row,
+            BrAPIObservation obs,
+            boolean includeTimestamp,
+            BrAPIObservationVariable var,
+            Program program) {
+        String varName = Utilities.removeProgramKey(obs.getObservationVariableName(), program.getKey());
+        if (var.getScale().getDataType().equals(BrAPITraitDataType.ORDINAL)) {
+            row.put(varName, Integer.parseInt(obs.getValue()));
```

Comment:
I think this has to be set as a `String` type for the same reason specified on this [comment](https://github.com/Breeding-Insight/bi-api/pull/259/#discussion_r1220199475)

Reply:
yep, done.